### PR TITLE
Use hwloc instead of libnuma for os_provider

### PR DESCRIPTION
**Pull request by igchor**
This allows us to get rid of libnuma as a dependency.

One missing features from hwloc is lack of support for: `UMF_NUMA_FLAGS_MOVE_ALL`. Do we care about this?"
Also, I think we need more tests before we can consider merging this. Something where we actually verify if memory is bound to a specific NUAMA node.

### CODE REVIEW

**Comment by igchor** [comment]
CI fails because I don't know how to best install hwloc on windows and macos yet ;d

**Comment by ldorau** [comment]
> CI fails because I don't know how to best install hwloc on windows and macos yet ;d

I think we can keep OS provider disabled on Windows and MacOS.

**Comment by bratpiorka** [comment]
@igchor I'm not sure about this change. WinAPI already supports NUMA (https://learn.microsoft.com/en-us/windows/win32/procthread/numa-support#numa-api) and with your changes, the UMF would require an additional dependency. Maybe we could use hwloc instead of libnuma only for Linux?

**Comment by igchor** [comment]
> @igchor I'm not sure about this change. WinAPI already supports NUMA (https://learn.microsoft.com/en-us/windows/win32/procthread/numa-support#numa-api) and with your changes, the UMF would require an additional dependency. Maybe we could use hwloc instead of libnuma only for Linux?

I thought we'll need to rely on hwloc anyway for topology discovery (e.g. to implement HOST memspace).

**Comment by igchor** [comment]
> > CI fails because I don't know how to best install hwloc on windows and macos yet ;d
> 
> I think we can keep OS provider disabled on Windows and MacOS as it is currently for now.

Ok. Done.

**Comment by bratpiorka** [comment]
> > @igchor I'm not sure about this change. WinAPI already supports NUMA (https://learn.microsoft.com/en-us/windows/win32/procthread/numa-support#numa-api) and with your changes, the UMF would require an additional dependency. Maybe we could use hwloc instead of libnuma only for Linux?
> 
> I thought we'll need to rely on hwloc anyway for topology discovery (e.g. to implement HOST memspace).

OK but what I'm thinking about is that installing hwloc on Windows could be not as easy as for linux. I think we should add short description to the readme how to do that. Also, we should think which hwloc version we would require. Recently hwloc team released 2.10 version with some nice features supporting heterogeneous memory. Maybe for a stand-alone version (without oneAPI base toolkit) we should ship hwloc binaries inside deps or something.

**Comment by igchor** [comment]
> > > @igchor I'm not sure about this change. WinAPI already supports NUMA (https://learn.microsoft.com/en-us/windows/win32/procthread/numa-support#numa-api) and with your changes, the UMF would require an additional dependency. Maybe we could use hwloc instead of libnuma only for Linux?
> > 
> > 
> > I thought we'll need to rely on hwloc anyway for topology discovery (e.g. to implement HOST memspace).
> 
> OK but what I'm thinking about is that installing hwloc on Windows could be not as easy as for linux. I think we should add short description to the readme how to do that. Also, we should think which hwloc version we would require. Recently hwloc team released 2.10 version with some nice features supporting heterogeneous memory. Maybe for a stand-alone version (without oneAPI base toolkit) we should ship hwloc binaries inside deps or something.

For now, the OS provider is not implemented for Windows anyway - when we do implement it, then yes, I agree, we should extend the readme. The version of hwloc we need will probably depend on the discovery use-case. For OS provider, we could even use 2.0

**Comment by vinser52** [comment]
> Maybe for a stand-alone version (without oneAPI base toolkit) we should ship hwloc binaries inside deps or something.

@bratpiorka what do you mean by "standalone version"? We never discussed the distribution model for UMF except as part of oneAPI or source code on GitHub?

**Comment by bratpiorka** [comment]
> > Maybe for a stand-alone version (without oneAPI base toolkit) we should ship hwloc binaries inside deps or something.
> 
> @bratpiorka what do you mean by "standalone version"? We never discussed the distribution model for UMF except as part of oneAPI or source code on GitHub?

Ok I was thinking about the far future. Sergey - do you know is hwloc lib is added into oneAPI?

**Comment by vinser52** [comment]
> Sergey - do you know is hwloc lib is added into oneAPI?

Yes, it was added for TCM needs.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -10,17 +10,16 @@ set(UMF_SOURCES
     memory_pool.c
     memory_provider.c
     memory_provider_get_last_failed.c
-    provider/provider_os_memory.c
```
Comment (bratpiorka) [inline src/CMakeLists.txt:13]: please do not change any windows - related code here

File: .github/workflows/basic.yml [hunk 2]
```diff
@@ -33,7 +33,7 @@ jobs:
       - name: Install apt packages
         run: |
           sudo apt-get update
-          sudo apt-get install -y clang cmake libnuma-dev libjemalloc-dev libtbb-dev
```
Comment (bratpiorka) [inline .github/workflows/basic.yml:36]: please update the README

**Review by bratpiorka — CHANGES_REQUESTED** [review]

**Comment by bratpiorka** [comment]
please rebase - also I think we could set this PR to "ready to review"

File: src/CMakeLists.txt (hunk 1 above)
Comment (igchor) [reply src/CMakeLists.txt:13]: Why? I don't see any reason for keeping files that just have stubs. 

Also, this PR makes it necessary to store extra state (hwloc_topology) in OS provider struct. It's easier to keep it in an OS-agnostic way vs defining a struct that is OS-specific. We will probably move the hwloc_topology to another place anyway in future (to some library ctor or `initialize` method so we can support discovery).

Comment (bratpiorka) [reply src/CMakeLists.txt:13]: ok let's get back to this when we start adding support for windows

**Comment by igchor** [comment]
I've rebased this PR and verified it works fine with tests from: https://github.com/oneapi-src/unified-memory-framework/pull/160 I think we can consider mergining it now. There are still some tests that would be useful (e.g. https://github.com/oneapi-src/unified-memory-framework/pull/144) but I don't think we need to actually wait on those.

File: .github/workflows/basic.yml [hunk 3]
```diff
@@ -81,7 +81,7 @@ jobs:
...
+          sudo apt-get install -y clang cmake libhwloc-dev libnuma-dev libjemalloc-dev libtbb-dev
```
Comment (PatKamin) [inline .github/workflows/basic.yml:84]: `linuma-dev` can be removed now?

Comment (igchor) [reply .github/workflows/basic.yml:84]: It's still needed in tests. We could use hwloc in tests as well but I think using numa gives us more confidence that the integration is done correctly.

**Review by bratpiorka — APPROVED** [review]

**Review by PatKamin — APPROVED** [review]

**Pull request merged**